src/OPTokenClaim.sol

```diff
@@ -54,62 +70,146 @@ contract OPTokenClaim is Ownable {
+    }
+
+    /// claim subscribed OP reward
+    function claimOP(address account) external {
+        Config memory _config = config;
+
+        // users claim reward for last epoch, so claims start at epoch 1
+        uint256 epochNum = _currentEpoch(_config);
+        require(epochNum > 0, "claims have not started yet");
+
+        uint256 lastEpochNum;
+        unchecked {
+            lastEpochNum = epochNum - 1;
+        }
+        require(lastEpochNum < _config.maxEpoch, "claims ended");
 
-        require(
-            !epochToAddressClaimed[epoch][account],
-            "already claimed for this epoch"
-        );
+        // check if subscribed and if already claimed
+        require(epochToSubscribedEXP[lastEpochNum][account] > 0, "didn't subscribe or already claimed");
 
-        uint256 claimableOP = _calcReward(account);
+        uint256 OPReward = _calcReward(account, lastEpochNum);
 
         // set claimed to true
-        epochToAddressClaimed[epoch][account] = true;
+        epochToSubscribedEXP[lastEpochNum][account] = 0;
+
+        // subscribe for next epoch
+        if (epochNum < _config.maxEpoch) {
+            subscribe(account);
+        }
 
         // transfer OP to account
-        require(
-            OP.transferFrom(treasury, account, claimableOP),
-            "Transfer failed"
-        );
+        require(OP.transferFrom(treasury, account, OPReward), "Transfer failed");
 
-        emit OPClaimed(account, epoch, claimableOP);
+        emit OPClaimed(account, lastEpochNum, OPReward);
     }
 
     /// @dev reverts if claims have not started yet
-    function currentEpoch(Config memory _config)
-        internal
-        view
-        returns (uint256 epochNumber)
-    {
-        require(block.timestamp >= _config.start, "claim period not started");
+    function _currentEpoch(Config memory _config) internal returns (uint256 epochNumber) {
+        require(block.timestamp >= _config.start, "reward dist not started yet");
         unchecked {
             epochNumber = (block.timestamp - _config.start) / 30 days;
         }
+
+        // push new epoch to array if needed
+        while ((epochs.length - 1) < epochNumber) {
+            epochs.push();
+        }
     }
 
-    function _calcReward(address account) internal view returns (uint256) {
-        uint256 expBalance = EXP.balanceOf(account);
-        require(expBalance > 0, "address has no exp");
+    // calculates reward of account for given epoch
+    function _calcReward(address account, uint256 epochNum) internal view returns (uint256) {
+        // calculate the total reward of given epoch
+        EpochInfo storage epoch = epochs[epochNum];
+        uint256 totalEXP = epoch.totalEXP;
+        uint256 subscribedAccounts = epoch.numAccounts;
 
-        // cap at 99 EXP
-        if (expBalance > 99 ether) {
-            expBalance = 99 ether;
-        }
+        unchecked {
+            uint256 totalReward = totalEXP * 5 - 4 ether * subscribedAccounts;
```

Comment:
looking closer at the math, I think there's a mistake here 

`(totalEXP * 5) - (4 ether * subscribedAccounts)`

vs 

`(totalEXP * 5 - 4 ether) * subscribedAccounts`

cc @beskay, let's wait to hear from @Dhannte to see if we can simplify the formula. I'd be in favor of getting rid of the `- 4` thing entirely

Reply:
The first equation is correct, for example, we have 21 subscribed accounts with 99 EXP each:
totalEXP = 99 * 21 = 2079 EXP
total OP we have to distribute: (99 * 5 - 4) * 21 = 10311 OP

Above equation:
(2079 * 5) - 4 * 21 = 10311 OP

Im fine with simplifying the formula though.